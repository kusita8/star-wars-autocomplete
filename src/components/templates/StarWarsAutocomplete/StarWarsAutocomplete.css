.starWarsAutocomplete {
  width: 100%;
}

.starWarsAutocomplete > div {
  margin-bottom: 4rem;
}
