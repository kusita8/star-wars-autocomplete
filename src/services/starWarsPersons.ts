import { ApiResponse } from "../types/ApiResponse.type";
import { Person } from "../types/Person.interface";

export const getStarWarsPersons = async (name?: string): Promise<ApiResponse<Person[]>> => {
  const data = await fetch(`https://swapi.dev/api/people${name ? `?search=${name}` : ""}`);
  return data.json();
};
