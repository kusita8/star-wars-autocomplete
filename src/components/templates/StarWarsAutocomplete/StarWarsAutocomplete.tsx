import { ChangeEvent, useRef } from "react";
import { useStarWarsData } from "../../../hooks/useStarWarsData";
import { debounce } from "../../../utils/debounce";
import { Input } from "../../atoms/Input/Input";
import { List } from "../../molecules/List/List";
import "./StarWarsAutocomplete.css";

export const StarWarsAutocomplete = () => {
  const { data, loading, error, fetchData } = useStarWarsData();
  const inputRef = useRef<HTMLInputElement>(null);

  const handleSearch = debounce(async (e: ChangeEvent<HTMLInputElement>) => {
    await fetchData(e.target.value);
  }, 300);

  return (
    <div className="starWarsAutocomplete">
      <Input ref={inputRef} label="Name" onChange={handleSearch} />
      {!error && <List search={inputRef.current?.value} persons={data} loading={loading} />}
    </div>
  );
};
