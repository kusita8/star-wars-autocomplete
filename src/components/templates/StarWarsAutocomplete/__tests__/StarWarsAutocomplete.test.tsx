import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { StarWarsAutocomplete } from "../StarWarsAutocomplete";
import * as service from "../../../../services/starWarsPersons";
import { getStarWarsPersons } from "../../../../services/__mocks__/starWarsPersons";

describe("StarWarsAutocomplete", () => {
  test("it should match the snapshot", async () => {
    const { asFragment } = render(<StarWarsAutocomplete />);

    await waitFor(() => {
      expect(screen.getByText("Luke Skywalker")).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test("it should search persons and highlight text", async () => {
    const mockGetStarWarsPersons = jest.fn(getStarWarsPersons);
    const spy = jest
      .spyOn(service, "getStarWarsPersons")
      .mockImplementation(mockGetStarWarsPersons);

    const { container } = render(<StarWarsAutocomplete />);

    await waitFor(() => {
      expect(screen.getByText("Luke Skywalker")).toBeInTheDocument();
    });

    fireEvent.change(container.querySelector("input") as HTMLInputElement, {
      target: {
        value: "luke",
      },
    });

    await waitFor(() => {
      expect(mockGetStarWarsPersons).toHaveBeenCalledWith("luke");
      expect(container.querySelectorAll(".listItem").length).toBe(1);
    });

    expect(container.querySelectorAll(".listItem__highlight")[0].innerHTML).toBe("L");
    expect(container.querySelectorAll(".listItem__highlight").length).toBe(8);

    spy.mockRestore();
  });

  test("it should show not found message when there isn't a match", async () => {
    const { container } = render(<StarWarsAutocomplete />);

    await waitFor(() => {
      expect(screen.getByText("Luke Skywalker")).toBeInTheDocument();
    });

    fireEvent.change(container.querySelector("input") as HTMLInputElement, {
      target: {
        value: "not a person",
      },
    });

    await waitFor(() => {
      expect(screen.getByText("No results found for not a person")).toBeInTheDocument();
    });
  });

  test("it should display alert when there is an error", async () => {
    const mockAlert = jest.fn();
    const spy1 = jest.spyOn(window, "alert").mockImplementation(mockAlert);
    const spy2 = jest.spyOn(service, "getStarWarsPersons").mockImplementation(() => {
      throw new Error("");
    });

    render(<StarWarsAutocomplete />);

    await waitFor(() => {
      expect(mockAlert).toHaveBeenCalledWith("An error occurred, please try again later");
    });

    spy1.mockRestore();
    spy2.mockRestore();
  });
});
