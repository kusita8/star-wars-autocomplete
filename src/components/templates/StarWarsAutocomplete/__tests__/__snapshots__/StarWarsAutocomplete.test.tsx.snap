// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarWarsAutocomplete it should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="starWarsAutocomplete"
  >
    <div>
      <label>
        Name
      </label>
      <input
        class="input"
        type="text"
      />
    </div>
    <div
      class="list"
    >
      <a
        class="listItem"
        href="https://swapi.dev/api/people/1/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Luke Skywalker
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/2/"
        rel="noopener noreferrer"
        target="_blank"
      >
        C-3PO
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/3/"
        rel="noopener noreferrer"
        target="_blank"
      >
        R2-D2
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/4/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Darth Vader
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/5/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Leia Organa
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/6/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Owen Lars
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/7/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Beru Whitesun lars
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/8/"
        rel="noopener noreferrer"
        target="_blank"
      >
        R5-D4
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/9/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Biggs Darklighter
      </a>
      <a
        class="listItem"
        href="https://swapi.dev/api/people/10/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Obi-Wan Kenobi
      </a>
    </div>
  </div>
</DocumentFragment>
`;
